File: app/Services/WordPress/WordPressTunnel.php
<?php

namespace App\Services\WordPress;

use Illuminate\Process\InvokedProcess;
use Illuminate\Support\Facades\Process;
use RuntimeException;

/**
 * Opens/closes the SSH tunnel `scripts/wordpress_tunnel.py` needs to reach
 * the WordPress/WooCommerce site's MySQL database — that database only
 * binds to localhost on its own host (confirmed live: a direct connection
 * from this app's network is rejected at the MySQL host-ACL level), so it's
 * only reachable by tunneling through SSH first, the same way the site's
 * own Navicat client is configured.
 *
 * The tunnel itself is a standalone Python/paramiko script rather than pure
 * PHP — phpseclib (the closest PHP SSH library) has no first-class local
 * port-forward API to build this on top of with any confidence, while the
 * Python script is the exact mechanism already proven working live against
 * this database this session.
 */
class WordPressTunnel
{
    private const LOCAL_PORT = 33061;

    private const READY_TIMEOUT_SECONDS = 15;

    private ?InvokedProcess $process = null;

    public function open(): void
    {
        $config = config('services.wordpress_db');

        if (empty($config['ssh_host']) || empty($config['ssh_username']) || empty($config['ssh_password']) || empty($config['db_database'])) {
            throw new RuntimeException('WordPress DB tunnel is not configured — set WORDPRESS_SSH_HOST/WORDPRESS_SSH_USERNAME/WORDPRESS_SSH_PASSWORD/WORDPRESS_DB_DATABASE (and the other WORDPRESS_DB_*/WORDPRESS_SSH_* vars) in .env.');
        }

        $scriptPath = base_path('scripts/wordpress_tunnel.py');

        // Credentials go through env(), never as command-line arguments, so
        // they never show up in a process listing (ps/tasklist).
        $this->process = Process::env([
            'WP_SSH_HOST' => $config['ssh_host'],
            'WP_SSH_PORT' => (string) $config['ssh_port'],
            'WP_SSH_USERNAME' => $config['ssh_username'],
            'WP_SSH_PASSWORD' => $config['ssh_password'],
            'WP_DB_HOST' => $config['db_host'],
            'WP_DB_PORT' => (string) $config['db_port'],
            'WP_LOCAL_PORT' => (string) self::LOCAL_PORT,
        ])
            // The tunnel must outlive the whole sync run, not Symfony
            // Process's own 60s default — this caps it well above any
            // realistic run instead of disabling it outright.
            ->timeout(3600)
            ->start(['python3', $scriptPath]);

        $deadline = microtime(true) + self::READY_TIMEOUT_SECONDS;
        $buffer = '';

        while (microtime(true) < $deadline) {
            if (!$this->process->running()) {
                $error = trim($this->process->errorOutput().$this->process->output());
                throw new RuntimeException("WordPress SSH tunnel process exited before becoming ready: {$error}");
            }

            $buffer .= $this->process->latestOutput();
            if (str_contains($buffer, 'TUNNEL READY')) {
                return;
            }

            usleep(200_000);
        }

        $this->process->stop();
        throw new RuntimeException('Timed out waiting for the WordPress SSH tunnel to become ready.');
    }

    public function close(): void
    {
        $this->process?->stop();
        $this->process = null;
    }

    public function localPort(): int
    {
        return self::LOCAL_PORT;
    }
}

File: scripts/wordpress_tunnel.py
r"""
Opens an SSH-forwarded local TCP port to the WordPress/WooCommerce site's
MySQL database (only reachable via SSH tunnel — confirmed live that a
direct connection from this app's network is rejected at the MySQL host-ACL
level, matching how the site's own Navicat client is configured).

Invoked by App\Services\WordPress\WordPressTunnel — never run manually with
credentials on the command line; all connection details come from
environment variables so they never appear in a process listing:

  WP_SSH_HOST, WP_SSH_PORT, WP_SSH_USERNAME, WP_SSH_PASSWORD
  WP_DB_HOST, WP_DB_PORT           (the DB's address as seen FROM the SSH
                                     host — typically "localhost"/3306,
                                     since MySQL there only binds to loopback)
  WP_LOCAL_PORT                     (local port to bind the tunnel entry on)

Prints "TUNNEL READY" to stdout once the local port is accepting
connections, then blocks until killed (SIGTERM/SIGINT) so the parent
process can treat that line as the readiness signal.

Cleanup note: the SIGTERM handler below only actually runs when the parent
kills this process on POSIX. On Windows, Symfony Process (WordPressTunnel's
caller) always force-kills via `taskkill /F /T` regardless of what signal
was requested — this process gets no chance to run its own `finally` block
there. That's fine correctness-wise (the OS closes this process's sockets,
including the local tunnel port, the instant it's killed, whether or not
our own cleanup code ran) — it just means "close the SSH session politely"
never happens on Windows, only "the connection drops."
"""

import os
import signal
import socket
import sys
import threading
import time

import paramiko


class ForwardServer(threading.Thread):
    def __init__(self, transport, local_port, remote_host, remote_port):
        super().__init__(daemon=True)
        self.transport = transport
        self.remote_host = remote_host
        self.remote_port = remote_port
        self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        try:
            self.sock.bind(("127.0.0.1", local_port))
            self.sock.listen(5)
        except OSError as e:
            # Most likely a previous run's tunnel is still holding this
            # port — surfaced as one clear line instead of a raw traceback,
            # so WordPressTunnel::open() (which reads this process's
            # stderr into its RuntimeException message) reports something
            # actionable.
            raise RuntimeError(f"Could not bind local port {local_port}: {e}") from e
        self.running = True

    def run(self):
        while self.running:
            try:
                client_sock, _addr = self.sock.accept()
            except OSError:
                break
            threading.Thread(target=self._handle, args=(client_sock,), daemon=True).start()

    def _handle(self, client_sock):
        try:
            chan = self.transport.open_channel(
                "direct-tcpip", (self.remote_host, self.remote_port), client_sock.getpeername()
            )
        except Exception as e:
            print(f"Channel open failed: {e}", file=sys.stderr)
            client_sock.close()
            return

        def pipe(src, dst):
            try:
                while True:
                    data = src.recv(4096)
                    if not data:
                        break
                    dst.sendall(data)
            except Exception:
                pass
            finally:
                for s in (src, dst):
                    try:
                        s.close()
                    except Exception:
                        pass

        threading.Thread(target=pipe, args=(client_sock, chan), daemon=True).start()
        threading.Thread(target=pipe, args=(chan, client_sock), daemon=True).start()

    def stop(self):
        self.running = False
        try:
            self.sock.close()
        except Exception:
            pass


def main():
    ssh_host = os.environ["WP_SSH_HOST"]
    ssh_port = int(os.environ.get("WP_SSH_PORT", "22"))
    ssh_username = os.environ["WP_SSH_USERNAME"]
    ssh_password = os.environ["WP_SSH_PASSWORD"]
    db_host = os.environ.get("WP_DB_HOST", "localhost")
    db_port = int(os.environ.get("WP_DB_PORT", "3306"))
    local_port = int(os.environ["WP_LOCAL_PORT"])

    client = paramiko.SSHClient()
    client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
    client.connect(ssh_host, port=ssh_port, username=ssh_username, password=ssh_password, timeout=10)
    transport = client.get_transport()

    try:
        forwarder = ForwardServer(transport, local_port, db_host, db_port)
    except RuntimeError as e:
        print(str(e), file=sys.stderr, flush=True)
        client.close()
        sys.exit(1)

    forwarder.start()
    time.sleep(0.3)

    stop_event = threading.Event()

    def handle_signal(signum, frame):
        stop_event.set()

    signal.signal(signal.SIGTERM, handle_signal)
    signal.signal(signal.SIGINT, handle_signal)

    print("TUNNEL READY", flush=True)

    try:
        while not stop_event.is_set():
            time.sleep(0.5)
    finally:
        forwarder.stop()
        client.close()


if __name__ == "__main__":
    main()
